services/gitdiff/gitdiff.go

```diff
@@ -1497,6 +1514,66 @@ func GetDiff(gitRepo *git.Repository, opts *DiffOptions, files ...string) (*Diff
 	return diff, nil
 }
 
+// GetUserSpecificDiff is like GetDiff, except that user specific data such as which files the given user has already viewed on the given PR will also be set
+func GetUserSpecificDiff(userID int64, pull *models.PullRequest, gitRepo *git.Repository, opts *DiffOptions, files ...string) (*Diff, error) {
+	diff, err := GetDiff(gitRepo, opts, files...)
+	if err != nil {
+		return nil, err
+	}
+	review, err := pulls.GetNewestReview(userID, pull.ID)
+	if err != nil || review == nil || review.UpdatedFiles == nil {
+		return diff, err
+	}
+
+	// Prepation to check for each file whether it was changed since the last review
+	changedFiles, err := gitRepo.GetFilesChangedBetween(review.CommitSHA, pull.HeadBranch)
+	if err != nil {
+		return diff, err
+	}
+
+	filesChangedSinceLastDiff := make(map[string]pulls.ViewedState)
+outer:
+	for _, diffFile := range diff.Files {
+		fileViewedState := review.UpdatedFiles[diffFile.GetDiffFileName()]
+
+		// Check whether it was previously checked whether the file has changed since the last review
+		if fileViewedState == pulls.HasChanged {
+			diffFile.HasChangedSinceLastReview = true
+			continue
+		}
+
+		filename := diffFile.GetDiffFileName()
+
+		// Check explicitly whether the file has changed since the last review
+		for _, changedFile := range changedFiles {
+			diffFile.HasChangedSinceLastReview = filename == changedFile
+			if diffFile.HasChangedSinceLastReview {
+				filesChangedSinceLastDiff[filename] = pulls.HasChanged
+				continue outer // We don't want to check if the file is viewed here as that would fold the file, which is in this case unwanted
+			}
+		}
+		// Check whether the file has already been viewed
+		if fileViewedState == pulls.Viewed {
+			diffFile.IsViewed = true
+			diff.NumViewedFiles++
+		}
+	}
+
+	// Explicitly store files that have changed in the database, if any is present at all.
+	// This has the benefit that the "Has Changed" attribute will be present as long as the user does not explicitly mark this file as viewed, so it will even survive a page reload after marking another file as viewed.
+	// On the other hand, this means that even if a commit reverting an unseen change is committed, the file will still be seen as changed.
+	if len(filesChangedSinceLastDiff) > 0 {
+		go func() {
+			err := pulls.UpdateReview(review.UserID, review.PullID, review.CommitSHA, filesChangedSinceLastDiff)
+			if err != nil {
+				log.Warn("Could not update review for user %d, pull %d, commit %s and the changed files %v: %v", review.UserID, review.PullID, review.CommitSHA, filesChangedSinceLastDiff, err)
+			}
+		}()
```

Comment:
Using goroutine here seems not a good idea. If unnecessary, just do the update directly.

Reply:
https://github.com/go-gitea/gitea/pull/19007#discussion_r835738315

TLDR: It simply improves performance. There's no need to wait for it to complete, and it is a quite expensive operation.

Reply:
Will add a comment.

Reply:
I think you shouldn't do it. Some questions about such design:
1. Why it is so slow to take 300ms? Can't it be optimized?
2. The goroutine loses the context. Gitea team are working to make sure every goroutine run in correct context recently, so I do not think we should not introduce new breakings. If you have time, you can help to make all new functions accept a `Context` argument. See https://github.com/go-gitea/gitea/pull/19163 and zeripath's plan (https://github.com/go-gitea/gitea/issues/18345#issuecomment-1017714851) for more details 
3. People may not expect the `GetUserSpecificDiff` do some updates inside .... it's really a surprise, indeed the function name is called `GetXxx`. I think there could be a better name, or move the update code out of the `GetXxx` function.

Reply:
> Why it is so slow to take 300ms? Can't it be optimized?

Well, if you tell me how, I'll gladly optimize it.
My guess is that the branch `previousReview, err := getNewestReviewApartFrom(...)` inside this method is the reason for this performance drop, but I have no idea how to optimize that...
I mean, it is literally only a `Get` with a few conditions. There is not even a JOIN...

> People may not expect the `GetUserSpecificDiff` do some updates inside

Absolutely agree. But I couldn't think of a better way to handle this case which, as the comment indicates, must be done at some point.
I can only see three ways, each of them has disadvantages:
1. Extracting the `Has Changed`- logic into a second method, e.g. `checkHasChanged`. Disadvantage: requires both callers to not forget to call the function and an additional loop. Advantage: It would make more sense to pass a Ctx only to this function instead of to `GetUserSpecificDiff`.
1. returning a third return value, a list of files that should be marked as changed. Disadvantage: This return value seems confusing, and each caller must ensure to update the entry in the database.
1. Update inside `GetUserSpecificDiff`. Current approach. Disadvantages: users don't expect this function to update something, ctx parameter seems out of place.

Reply:
> My guess is that the branch ``

Could you provide more details about how did you do the benchmark? I have left another comment about the database index. 

> About the name of `GetUserSpecificDiff`

If you prefer to do update inside it, then it can have a new name, like `SyncUserSpecificDiff`

Reply:
The benchmark was
1. starting a local Gitea instance without db
2. creating a user and PR
3. open the browser console
4. in another tab, add and edit files
5. mark the files viewed in order to test all updating states
6. measure how long it takes until the browser received the status code

Your hint seems to have worked already: Instead of timings in the range of `250-290ms`, I now get something like `150-180ms`.

Reply:
Okay, interesting. I just added **a lot** of timing statements, the thing that is so slow here is the actual xorm `Update` statement...
How do I optimize an update statement?

Reply:
Are you sure your benchmark is reasonable and correct? For a RDMS, a simple update would succeed in less than 1ms. You should have a full understanding about your environment and your benchmark first.

What kind database? What's the table structure? How much data? What kind disk? How much memory? What are the database settings? Even how about the CPU usage? How about network packet loss and latency? How about other SQL queries? What does `explain` say? Define and clarify problems first before take actions.

Reply:
command: `TAGS="sqlite sqlite_unlock_notify" make watch`
db: `SQLite`
db settings:
```text
[database]
DB_TYPE  = sqlite3
HOST     = 127.0.0.1:3306
NAME     = gitea
USER     = gitea
PASSWD   =
SCHEMA   =
SSL_MODE = disable
CHARSET  = utf8
PATH     = /home/leon/git/gitea/data/gitea.db
LOG_SQL  = false
```
disk: `hdd`
table structure: Freshly generated database created through the migrations.
run mode: `prod`
RAM and CPU usage: no clue.
---
I've never mentioned that it will be like this for every user, that's simply what I encounter while testing (and implementing) the feature.

Reply:
> The goroutine loses the context. Gitea team are working to make sure every goroutine run in correct context recently

Could you please tell me how to add the context to the go routine?
I can imagine what I need to do to pass it to the database, but how do I set the context for the goroutine?

Reply:
I mean do not use goroutine here.